go-types/filesource/filesource.go

```diff
@@ -0,0 +1,563 @@
+}
+
+func (r *reader) shouldSkip(obj ObjectInfo) (_ bool, reason string) {
+	// Have we already processed through this filename in this sweep?
+	if r.state.Path > obj.Path {
+		return true, "state.Path > obj.Path"
+	}
+	if r.state.Path == obj.Path && r.state.Complete {
+		return true, "state.Path == obj.Path && Complete"
+	}
+	// Is the path excluded by our regex ?
+	if r.pathRe != nil && !r.pathRe.MatchString(obj.Path) {
+		return true, "regex not matched"
+	}
+	// Is the path modified before our window (inclusive) ?
+	if r.state.MinMod.After(obj.ModTime) {
+		return true, "MinMod.After(obj.ModTime)"
+	}
+	// Is the path modified after our window (exclusive) ?
+	if r.state.MaxMod.Before(obj.ModTime) {
+		return true, "MaxMod.After(obj.ModTime)"
+	}
+
+	return false, ""
+}
+
+func (r *reader) processObject(ctx context.Context, obj ObjectInfo) error {
+	rr, obj, err := r.store.Read(ctx, obj)
+	if err != nil {
+		return err
+	}
+	defer rr.Close()
+
+	r.log("processing file %q", obj.Path)
+
+	tmp, err := ioutil.TempFile("", "parser-config-*.json")
+	if err != nil {
+		return fmt.Errorf("creating parser config: %w", err)
+	}
+	defer os.Remove(tmp.Name())
+
+	if err = r.makeParseConfig(obj).WriteToFile(tmp); err != nil {
+		return fmt.Errorf("writing parser config: %w", err)
+	}
+
+	// Skip tracks records remaining to skip, because they've already
+	// been ingested in a prior invocation of this capture.
+	var skip int
+
+	if r.state.Path == obj.Path {
+		if r.state.Complete {
+			panic("should have been filtered already")
+		}
+		skip = r.state.Records
+	}
+
+	r.state.Path = obj.Path
+	r.state.Records = 0
+	r.state.Complete = false
+
+	err = parser.ParseStream(ctx, tmp.Name(), rr, func(lines []json.RawMessage) error {
+		r.state.Records += len(lines)
+
+		if ll := len(lines); ll < skip {
+			skip -= ll
+			return nil
+		} else if skip != 0 {
+			lines = lines[skip:]
+			skip = 0
+		}
+
+		return r.commitLines(lines)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Write a final checkpoint to mark the completion of the file.
+	r.state.Complete = true
+	r.state.Records = 0
+
+	if err := r.commitLines(nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reader) makeParseConfig(obj ObjectInfo) *parser.Config {
+	var cfg = new(parser.Config)
+	if c := r.config.ParserConfig(); c != nil {
+		*cfg = c.Copy()
+	}
+
+	cfg.Filename = obj.Path
+	if obj.ContentType != "" {
+		cfg.ContentType = obj.ContentType
+	}
+	if obj.ContentEncoding != "" {
+		cfg.ContentEncoding = obj.ContentEncoding
+	}
+	// If the user supplied a location for this, then we'll use that. Otherwise, use the default
+	if cfg.AddRecordOffset == "" {
+		cfg.AddRecordOffset = metaOffsetLocation
+	}
+	if cfg.AddValues == nil {
+		cfg.AddValues = make(map[parser.JsonPointer]interface{})
+	}
+	if cfg.Schema == nil {
+		cfg.Schema = r.schema
+	}
+	cfg.Projections = r.projections
+
+	cfg.AddValues[metaFileLocation] = obj.Path
```

Comment:
It seems a little weird to me that the file path and record offset are handled differently. It feels like they should either both be user-configurable or neither should be. Personally, my instinct at this point would be to say that neither needs to be user-configurable, but given our direct embedding of the parser json schema that may be impractical. But one idea that comes to mind would be to unconditionally overwrite `AddRecordOffset` with the hard-coded default. This is based on the assumption that our strategy of directly embedding the parser schema will soon be replaced by a bespoke schema that includes only a subset of the fields exposed by the parser config. I can't really think of a reason why users would need to customize either of these, though.

Reply:
Agreed. I don't believe I changed this behavior from how the `source-s3` connector previously did it.

My preference would also be to yank these out as things which can be configured.

Reply:
Yeah, I had that thought that I probably wrote it this way originally. Just didn't think that through at the time.

Reply:
To do this properly would require that rust, and not Go, defines `metaOffsetLocation`, right?

Then it could be removed as a parser configurable.